idearium-lib/lib/mq/connection.js

```diff
@@ -0,0 +1,174 @@
+'use strict';
+
+var url = require('url'),
+    EventEmitter = require('events').EventEmitter,
+    amqp = require('amqplib'),
+    debug = require('debug')('idearium-lib:mq-connection');
+
+class Connection extends EventEmitter {
+
+    /**
+     * Constructor function
+     * @param  {String} connectionString   Rabbitmq server url
+     * @param  {Object} options            Rabbitmq SSL certificates see http://www.squaremobius.net/amqp.node/ssl.html for more details
+     * @param  {Number} reconnectTimeout   Timeout (milliseconds) to reconnect to rabbitmq server. Defaults to 5000
+     */
+    constructor(connectionString, options = {}, reconnectTimeout = 5000) {
+
+        if (!connectionString) {
+            throw new Error('connectionString parameter is required');
+        }
+
+        // Init EventEmitter
+        super();
+
+        this.url = connectionString;
+        this.options = options;
+        this.reconnectTimeout = reconnectTimeout;
+        this.connection = null;
+        this.state = 'disconnected';
+
+        // Support SSL based connections
+        // https://help.compose.com/docs/rabbitmq-connecting-to-rabbitmq#section-node-and-rabbitmq
+        if (!this.options.servername) {
+            this.options.servername = url.parse(this.url).hostname;
+        }
+
+    }
+
+    /**
+     * Establish connection to rabbitmq and reconnects if connection failure occurs.
+     */
+    connect() {
+
+        if (this.state === 'connected' || this.state === 'connecting') {
+            return null;
+        }
+
+        this.state = 'connecting';
+
+        debug(`Attempting to connect at ${this.url}`);
+
+        return amqp.connect(this.url, this.options)
+            .then(conn => this.connected(conn))
+            .catch((err) => {
+
+                // This is a bit ugly, but when connecting to RabbitMQ (with SSL)
+                // older versions can throw some weird errors.
+                // This tidies up a really persistent one that seems to happen on every connect.
+                if (err.message && !err.message.startsWith('socket hang up')) {
+
+                    debug('Unable to connect!');
+                    this.logError(err);
+                    this.emit('error', err);
+
+                }
+
+                return this.reconnect();
+
+            });
+
+    }
+
+    /**
+     * Executed once connected with the connection.
+     * @param  {Object} conn A connection object.
+     * @return Void
+     */
+    connected(conn) {
+
+        this.connection = conn;
+
+        // handle disconnection error
+        conn.on('error', (err) => {
+
+            debug('Disconnected!');
+            this.logError(err);
+            this.emit('error', err);
+
+            return this.reconnect();
+
+        });
+
+        debug('Connected');
+
+        this.state = 'connected';
+        this.emit('connect');
```

Comment:
@allanchau, yeah, I noted that issue, but I can't work out where to put a `return` statement.

I know for a fact that the https://github.com/squaremo/amqp.node library suffers from the same thing.

I'm not quite sure if it is our code or theirs... perhaps you could take a quick look at where you think it might go?